### PR TITLE
Refactor receive stream buffering

### LINKED ISSUE #735: read_unordered can issue redundant flow control credit
Stream assembly only deduplicates buffered data when the number of buffered frames exceeds a critical threshold. Because `read_unordered` pulls frames directly out of the buffer and does not keep track of previously-read data, receiving the same stream frame multiple times can lead to arbitrarily large increases in flow control limits. A malicious or buggy peer could leverage this to perform a memory-exhaustion DoS on an application that uses `read_unordered`.

One approach for a fix is to maintain a `RangeSet` defining which portions of the stream have not yet been read when using unordered reads, just as the assembler's offset defines that for ordered reads. Redundant data could then be discarded for both flow control and application processing purposes.

### PR DESCRIPTION
**Pull request by Ralith**
Fixes #735. Should marginally improve performance for ordered reads due to isolation of the (now somewhat more) complex `RangeSet` operations to the unordered case where they're necessary, leaving simple cursor-based logic for ordered behavior.

### CODE REVIEW

File: quinn-proto/src/connection/assembler.rs [hunk 1]
```diff
@@ -164,16 +175,26 @@ impl Assembler {
...
-    /// Whether we've received all data below `final_offset`
-    ///
-    /// `final_offset` must be greater than the offset of any received data.
-    pub(crate) fn is_complete_at(&self, final_offset: u64) -> bool {
-        debug_assert!(final_offset >= self.limit());
-        (self.recvd.len() == 1 && self.recvd.iter().next().unwrap() == (0..final_offset))
-            || final_offset == 0
+    /// Whether all data prior to `self.limit()` has been read
+    pub(crate) fn is_fully_read(&self) -> bool {
+        if let State::Unordered { ref recvd } = self.state {
```
Comment (djc) [inline quinn-proto/src/connection/assembler.rs:197]: Hmm, maybe write this as a `match` expression instead? Would improve the symmetry and obviate the fall-through case (which I think is not very obvious here).

File: quinn-proto/src/connection/assembler.rs [hunk 2]
```diff
@@ -7,19 +7,14 @@ use crate::range_set::RangeSet;
...
-    /// Whether any unordered reads have been performed, making this assembler unusable for ordered
-    /// reads
-    unordered: bool,
     /// Number of bytes read. Equal to offset when `unordered` is false, may or may not be
     /// otherwise.
     bytes_read: u64,
-    /// Offsets received so far
-    recvd: RangeSet,
+    /// First offset we haven't received any data at or after
+    limit: u64,
```
Comment (djc) [inline quinn-proto/src/connection/assembler.rs:17]: I feel like `limit` isn't a great name for this. Any other ideas? `end`, `next`, `unseen`? (Of these options I think I like `next` best.)

File: quinn-proto/src/connection/assembler.rs [hunk 3]
```diff
@@ -64,23 +62,28 @@ impl Assembler {
             None => return false,
         };
 
+        let offset = match self.state {
+            State::Ordered { ref mut offset } => offset,
```
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs:66]: It's the newer pattern, but I disagree that it's a better one, and I don't think there's a consensus either way (hence no clippy lint). I prefer this approach because the type of `offset` is determined by its immediate context, rather than the type of whatever was passed in to the `match`. This makes the type more obvious, especially for larger matches and matches on variables where it isn't immediately apparent whether they contain a reference.

File: quinn-proto/src/connection/assembler.rs [hunk 4]
```diff
@@ -64,23 +62,28 @@ impl Assembler {
...
+            _ => panic!("cannot perform ordered reads following unordered reads on a stream"),
```
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs:67]: I'd rate silent misbehavior (e.g. returning `None`) as worse than panicking, but I don't see any reason not to return an error. I'll add that tonight.

File: quinn-proto/src/connection/assembler.rs (hunk 1 above)
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs:197]: Yeah, that's definitely clearer.

File: quinn-proto/src/connection/assembler.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (djc) [reply quinn-proto/src/connection/assembler.rs:66]: Hmm, ok.

**Comment by djc** [comment]
(Also very well-factored commit here. In terms of preventing unnecessary rebase efforts, maybe submit the first four commits separately so we can merge them immediately?)

File: quinn-proto/src/connection/assembler.rs (hunk 2 above)
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs]: I went with `end`, which had precedent elsewhere in related code. `next` could be confusing because it strictly has no bearing on what will actually be received next; data can be transmitted in any order so long as flow control limits are respected.

**Comment by Ralith** [comment]
Cherry-picked the simple stuff into https://github.com/djc/quinn/pull/770.

File: quinn-proto/src/connection/assembler.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs]: Done.

**Comment by Ralith** [comment]
Rebased to account for #776. Will need another after #777 is merged, or vis versa.

**Comment by Ralith** [comment]
Rebased.

**Comment by Ralith** [comment]
Got rid of a bunch of unnecessary complexity regarding `is_fully_read`, and found and fixed a significant bug in the process.

File: quinn-proto/src/connection/assembler.rs [hunk 5]
```diff
@@ -104,7 +104,12 @@ impl Assembler {
     fn defragment(&mut self) {
         let buffered = self.data.iter().map(|c| c.bytes.len()).sum::<usize>();
         let mut buffer = BytesMut::with_capacity(buffered);
-        let mut offset = self.offset;
+        let mut offset = self
```
Comment (djc) [inline quinn-proto/src/connection/assembler.rs:109]: Should we have a test for this?

File: quinn-proto/src/connection/assembler.rs [hunk 6]
```diff
@@ -383,4 +421,35 @@ mod test {
...
+        assert_eq!(x.read_unordered(), None);
+    }
+
+    #[test]
+    fn unordered_dedup() {
+        let mut x = Assembler::new();
+        x.insert(3, Bytes::from_static(b"def"));
+        assert_eq!(x.read_unordered(), Some((3, Bytes::from_static(b"def"))));
+        assert_eq!(x.read_unordered(), None);
+        x.insert(0, Bytes::from_static(b"abcdefghi"));
```
Comment (djc) [inline quinn-proto/src/connection/assembler.rs:446]: I think large chunks are better than small chunks. If we get a large chunk and we have a small chunk buffered which is a subset of the large chunk, should we keep the larger chunk instead of splitting it up?

File: quinn-proto/src/connection/assembler.rs [hunk 7]
```diff
@@ -383,4 +421,35 @@ mod test {
...
+        x.insert(0, Bytes::from_static(b"abc"));
+        assert_eq!(x.read_unordered(), Some((0, Bytes::from_static(b"abc"))));
+        assert_eq!(x.read_unordered(), None);
+        x.insert(3, Bytes::from_static(b"def"));
+        assert_eq!(x.read_unordered(), Some((3, Bytes::from_static(b"def"))));
+        assert_eq!(x.read_unordered(), None);
+    }
+
+    #[test]
+    fn unordered_dedup() {
```
Comment (djc) [inline quinn-proto/src/connection/assembler.rs:441]: I think this is testing "middle" and "tail" duplication but not "head" duplication. Should we have a test for that as well to make sure we cover all the edge cases here?

File: quinn-proto/src/range_set.rs [hunk 8]
```diff
@@ -244,6 +271,43 @@ impl<'a> DoubleEndedIterator for EltIter<'a> {
     }
 }
 
+pub struct InsertDrain<'a> {
+    set: &'a mut RangeSet,
```
Comment (djc) [inline quinn-proto/src/range_set.rs:275]: Would be good to have documentation here as to what is the semantic meaning of each of these fields and also probably just the semantics of this type.

File: quinn-proto/src/range_set.rs [hunk 9]
```diff
@@ -138,6 +138,33 @@ impl RangeSet {
         before || after
     }
 
+    /// Add a range to the set, returning the ranges it replaced
+    pub fn insert_drain(&mut self, mut range: Range<u64>) -> InsertDrain<'_> {
+        let pred = if let Some((prev_start, prev_end)) = self.pred(range.start) {
```
Comment (djc) [inline quinn-proto/src/range_set.rs:143]: Is there some way to prevent the 3-level cascading `else { None }` here? (Something with `Option::filter()`, I suppose.)

(Also, a little documentation on `succ()` and, in particular `pred()`, might help. I guess I haven't looked at this code in much detail before, but `pred` primarily invoked "predicate" for me rather than `predecessor`.)

Don't love this name. Would it make sense to just call this `insert()`? It looks like there's some duplication between `insert()` and `insert_drain()` currently, so if there's not too much overhead it might make sense to not maintain both.

File: quinn-proto/src/range_set.rs [hunk 10]
```diff
@@ -138,6 +138,33 @@ impl RangeSet {
         before || after
     }
 
+    /// Add a range to the set, returning the ranges it replaced
```
Comment (djc) [inline quinn-proto/src/range_set.rs:141]:
```suggestion
    /// Add a range to the set, returning the intersection of current ranges with the new one
```

File: quinn-proto/src/range_set.rs (hunk 9 above)
Comment (Ralith) [reply quinn-proto/src/range_set.rs:143]: > Is there some way to prevent the 3-level cascading else { None } here?

We could do this with combinators, but the main block of code here has side effects on the data structure, and I'm leery of hiding that. Still, was able to unnest by one level, which is something.

> documentation

:+1: 

> Don't love this name.

I agree the name isn't great. How's `replace`?

> Would it make sense to just call this insert()?

I'm reluctant because, at least syntatically, we go to a bunch of extra effort to produce an iterator. This did prompt me to reexamine and simplify `insert`, though, which makes keeping it that much more pleasant.

File: quinn-proto/src/range_set.rs (hunk 8 above)
Comment (Ralith) [reply quinn-proto/src/range_set.rs:275]: Good idea.

File: quinn-proto/src/connection/assembler.rs (hunk 6 above)
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs:446]: That would require random-accessing into the `BinaryHeap` somehow to replace the shorter pre-existing chunk, or deferring the deduplication until `read_unordered`-time, which increases memory use. Given that `defragment` will keep a lid on fragmentation overall, and what I estimate to be a quite low incidence of cases where perfect efficiency here would help much, I'm inclined to stick with the easy thing.

File: quinn-proto/src/connection/assembler.rs (hunk 7 above)
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs:449]: Added a few lines to exercise that too.

File: quinn-proto/src/connection/assembler.rs (hunk 6 above)
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs:455]: Thinking on this further, I suspect unordered mode should have a `BTreeMap<u64, Bytes>` rather than the `RangeSet` + `BinaryHeap` arrangement. Then we could optimize for efficient memory layout just as easily as we currently preserve `RangeSet`'s invariants. Removing the first item from a `BTreeMap` to read is awkward, but when we're traversing it multiple times per packet anyway...

I'd rather not block this PR on it, though, since this PR is still a direct improvement on the status quo.

File: quinn-proto/src/range_set.rs (hunk 9 above)
Comment (djc) [reply quinn-proto/src/range_set.rs]: Much better!

File: quinn-proto/src/connection/assembler.rs (hunk 6 above)
Comment (djc) [reply quinn-proto/src/connection/assembler.rs:455]: Sounds good.

**Comment by djc** [comment]
Nice work.
